### ISSUE TITLE
Allow selecting a device based on the name of a Physical Device

### ISSUE THREAD

**Issue by charles-lunarg** [charles-lunarg/vk-bootstrap#102]
Currently, there is no easy way to get the list of names of the available physical devices nor select a physical device based on that list.

Suggested improvements:
1. Create a function to query the names of the physical devices (allowing applications to present that list to the user
2. Add a 'select' like function to `vkb::PhysicalDeviceSelector` that returns a list of all available physical devices which meet the criteria. This lets apps get all the physical devices which meet the requirements.
3. Add a name as a criteria for selecting a physical device.

**Comment by playmer** [comment]
Regarding 2: Should partially suitable devices be returned? Or should they only be returned if no fully suitable devices be returned? I think I lean towards the second, since that would allow the user/engine more flexibility. Maybe it could be a bool parameter for the function?

**Comment by charles-lunarg** [comment]
Partially suitable devices are a poor solution for the 'I must have X, I would like Y' because there isn't a clear way to find out if Y is available.
I would like to not support that here, or at least make it more explicit like with a bool parameter. `get_suitable_devices(bool return_partially_suitable)` and the like.

**Comment by charles-lunarg** [comment]
FYI, a partially suitable physical devices are from the desire to support apps where a known base spec is required but there is support for optional features, like raytracing. However, the API doesn't make it at all clear which features a given physical device supports. Like, you can be sure the device supports the required features but absolutely no clue about the optional stuff. 
I'm not asking for a solution to this issue, but that I'm open to discussion & ideas on how to improve the situation.

**Comment by playmer** [comment]
Yeah I didn't even know partially suitable devices were a thing in the library until I started poking at the select function :P so I don't yet have an opinion or idea, but I'll try to think about it.

**Comment by playmer** [comment]
So I've been factoring stuff out of `get_suitable_devices(bool)` and `select()` and rather than having a function to return just the device names, maybe we should just make public the `PhysicalDeviceDesc` struct and a function to get all of those since I already have that as a private function since both need it?

**Comment by jackie-fuchs** [comment]
I am confused as to why all of this needs to be refactored out when the name could just be added as a string to the PhysicalDevice struct that is already returned.

**Comment by playmer** [comment]
Well I refactored it out more to reduce code duplication between the two functions. But that would work as well of course.

**Comment by playmer** [comment]
The other reason was to not bring in `std::string` as a dependency on the header.

**Comment by charles-lunarg** [comment]
Wait, isn't the name in `VkPhysicalDeviceProperties::deviceName`?
So really we just need to return a list of `vkb::PhysicalDevice`s that meet the criteria rather than the first.

**Comment by jackie-fuchs** [comment]
Oh right, we had the name already from the start. Then yes all we need is the impl that returns a vector of **vkb::PhysicalDevice**.

**Comment by playmer** [comment]
Ah, I missed the `VkPhysicalDeviceProperties` on the `vkb::PhysicalDevice`, sorry for my confusion.

**Comment by charles-lunarg** [comment]
I'd argue that you missed it because its not very well documented. Adding a function that returned a vector of `vkb::PhysicalDevices` makes it easier to applications to use some of the existing requirements checking while being able to add their own.

**Comment by jackie-fuchs** [comment]
That being said, the point of the `vkb::PhysicalDeviceSelector` is to filter through devices based on the criteria it is configured with and the returned `vkb::PhysicalDevice`'s should not be excluded from that process.

**Comment by playmer** [comment]
Right, agreed, I think we're on the same page there.
